### ISSUE TITLE
Strong type api method responses

### ISSUE THREAD

**Issue by bilby91** [slackapi/node-slack-sdk#509]
### Description

Hello, I'm continuing my migration for `botbuilder-slack` to the new 4.x release. Today I realized that the responses from method calls are not type dependent on the method invocation, response type for every api call is `Promise<WebAPICallResult>`. Is there any plan to support strong typed responses ? I can work on this with my previous typings.

### What type of issue is this? (place an `x` in one of the `[ ]`)
- [ ] bug
- [X] enhancement (feature request)
- [ ] question
- [ ] documentation related
- [ ] testing related
- [ ] discussion

### Requirements (place an `x` in each of the `[ ]`)
* [X] I've read and understood the [Contributing guidelines](https://github.com/slackapi/node-slack-sdk/blob/master/.github/contributing.md) and have done my best effort to follow them.
* [X] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).
* [X] I've searched for any related issues and avoided creating a duplicate issue.

**Comment by aoberoi** [comment]
@bilby91 I’m glad to see your migration continue!

We’re having a discussion about a few different ways to get richer types in another issue. Specifically, I mentioned exactly what you’re asking for in point (4) in [this comment](https://github.com/slackapi/node-slack-sdk/issues/496#comment-372803268). I’d really appreciate your participation in that discussion so we can potentially totally come up with the steps to get there and divide the work.

**Comment by neekolas** [comment]
I was running into this issue as well. I got fed up trying to shut the Typescript compiler up every time I made an API call and wound up crawling the [Object Type](https://api.slack.com/types) definition pages, using the examples to generate a JSON Schema, and using the schema to build Typescript runtime validators and interfaces.

I am doing the same thing for the events API, which works really nicely, but the web responses have been giving me trouble. Isssue is that there aren't enough samples for me to infer which fields are reliably going to be there and which ones aren't. Here's an example of the generated schema:
_Not all of these fields should really be required_
```
{
  "$schema": "http://json-schema.org/draft-06/schema#",
  "$ref": "#/definitions/File",
  "definitions": {
    "File": {
      "type": "object",
      "additionalProperties": false,
      "properties": {
        "channels": { "type": "array", "items": { "type": "string" } },
        "comments_count": { "type": "integer" },
        "created": { "type": "integer" },
        "display_as_bot": { "type": "boolean" },
        "edit_link": { "type": "string" },
        "editable": { "type": "boolean" },
        "external_type": { "type": "string" },
        "filetype": { "type": "string" },
        "groups": { "type": "array", "items": { "type": "string" } },
        "id": { "type": "string" },
        "ims": { "type": "array", "items": { "type": "string" } },
        "initial_comment": { "$ref": "#/definitions/InitialComment" },
        "is_external": { "type": "boolean" },
        "is_public": { "type": "boolean" },
        "is_starred": { "type": "boolean" },
        "lines": { "type": "integer" },
        "lines_more": { "type": "integer" },
        "mimetype": { "type": "string" },
        "mode": { "type": "string" },
        "name": { "type": "string" },
        "num_stars": { "type": "integer" },
        "permalink": { "type": "string" },
        "permalink_public": { "type": "string" },
        "pinned_to": { "type": "array", "items": { "type": "string" } },
        "pretty_type": { "type": "string" },
        "preview": { "type": "string" },
        "preview_highlight": { "type": "string" },
        "public_url_shared": { "type": "boolean" },
        "reactions": {
          "type": "array",
          "items": { "$ref": "#/definitions/Reaction" }
        },
        "size": { "type": "integer" },
        "thumb_160": { "type": "string" },
        "thumb_360": { "type": "string" },
        "thumb_360_gif": { "type": "string" },
        "thumb_360_h": { "type": "integer" },
        "thumb_360_w": { "type": "integer" },
        "thumb_480": { "type": "string" },
        "thumb_480_h": { "type": "integer" },
        "thumb_480_w": { "type": "integer" },
        "thumb_64": { "type": "string" },
        "thumb_80": { "type": "string" },
        "timestamp": { "type": "integer" },
        "title": { "type": "string" },
        "url_private": { "type": "string" },
        "url_private_download": { "type": "string" },
        "user": { "type": "string" },
        "username": { "type": "string" }
      },
      "required": [
        "channels",
        "comments_count",
        "created",
        "display_as_bot",
        "edit_link",
        "editable",
        "external_type",
        "filetype",
        "groups",
        "id",
        "ims",
        "initial_comment",
        "is_external",
        "is_public",
        "is_starred",
        "lines",
        "lines_more",
        "mimetype",
        "mode",
        "name",
        "num_stars",
        "permalink",
        "permalink_public",
        "pinned_to",
        "pretty_type",
        "preview",
        "preview_highlight",
        "public_url_shared",
        "reactions",
        "size",
        "thumb_160",
        "thumb_360",
        "thumb_360_gif",
        "thumb_360_h",
        "thumb_360_w",
        "thumb_480",
        "thumb_480_h",
        "thumb_480_w",
        "thumb_64",
        "thumb_80",
        "timestamp",
        "title",
        "url_private",
        "url_private_download",
        "user",
        "username"
      ],
      "title": "File"
    },
    "InitialComment": {
      "type": "object",
      "additionalProperties": false,
      "title": "InitialComment"
    },
    "Reaction": {
      "type": "object",
      "additionalProperties": false,
      "properties": {
        "count": { "type": "integer" },
        "name": { "type": "string" },
        "users": { "type": "array", "items": { "type": "string" } }
      },
      "required": ["count", "name", "users"],
      "title": "Reaction"
    }
  }
}

```
and here are the interfaces.
```

export interface File {
  id: string;
  created: number;
  timestamp: number;
  name: string;
  title: string;
  mimetype: string;
  filetype: string;
  pretty_type: string;
  user: string;
  mode: string;
  editable: boolean;
  is_external: boolean;
  external_type: string;
  username: string;
  size: number;
  url_private: string;
  url_private_download: string;
  thumb_64: string;
  thumb_80: string;
  thumb_360: string;
  thumb_360_gif: string;
  thumb_360_w: number;
  thumb_360_h: number;
  thumb_480: string;
  thumb_480_w: number;
  thumb_480_h: number;
  thumb_160: string;
  permalink: string;
  permalink_public: string;
  edit_link: string;
  preview: string;
  preview_highlight: string;
  lines: number;
  lines_more: number;
  is_public: boolean;
  public_url_shared: boolean;
  display_as_bot: boolean;
  channels: string[];
  groups: string[];
  ims: string[];
  initial_comment: InitialComment;
  num_stars: number;
  is_starred: boolean;
  pinned_to: string[];
  reactions: Reaction[];
  comments_count: number;
}

export interface InitialComment {}

export interface Reaction {
  name: string;
  count: number;
  users: string[];
}

```

I'd love to cut out the crawling step if someone has a better data source. Half of those pages don't even have valid JSON in them...sigh. Without more input data the whole things is more art than science, unfortunately.

I'm willing to trade a steady stream of up-to-date Typescript definitions and JSON schemas for API responses, and the Events API, if someone has a steady stream of example data to offer.

**Comment by bilby91** [comment]
Hello @aoberoi !!

I've been looking into the issues and PRs and it seems that there has been some progress on response typings :) 

Are they complete by this moment ?

**Comment by stevengill** [comment]
There is a PR open currently adding typing for web api responses. https://github.com/slackapi/node-slack-sdk/pull/1078. We are hoping to land these changes in the next few months.

**Comment by jonocairns** [comment]
Is the main reasoning here that the API runs at it's own version - thus the package version can not really guarantee the responses are absolutely correct?

**Comment by seratch** [comment]
Let me closet this issue as https://github.com/slackapi/node-slack-sdk/pull/1188 added response types.

The pre-released version 6.2.0-rc.0 including the response types is already available: https://www.npmjs.com/package/@slack/web-api/v/6.2.0-rc.0

We are going to release v6.2.0 before long!
